feat: add `join` method to `Url` class

Meetesh-Saini opened the pull request:
- added support for URL path joining with optional trailing slashes and multiple arguments.

<!-- Thank you for your contribution! -->

## Change Summary

<!-- Please give a short summary of the changes. -->
This PR implements a feature based on pydantic/pydantic#9794 to join URL path into the base URL. It uses the `join` method from the `url` crate.

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->
<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->
fix pydantic/pydantic#9794

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Documentation reflects the changes where applicable
* [ ] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

davidhewitt commented on src/url.rs:
```diff
@@ -190,6 +190,37 @@ impl PyUrl {
+                add_trailing_slash(&mut arg_str);
+            }
+            new_url = join_path(&new_url, arg_str)?;
+        }
```
I think rather than this implementation we can use [`Url::path_segments_mut`](https://docs.rs/url/latest/url/struct.Url.html#method.path_segments_mut) to push each segment in turn.

I think adding a trailing slash is then as simple as pushing `""` as a final step.

Meetesh-Saini replied:
I agree with you, but wouldn't this approach negate the additional features of `Url::join`, such as joining queries and fragments, and supporting path traversal?
```rust
let mut a = Url::parse("http://example.com/").unwrap();
println!("{}", a.join("a/b/../c").unwrap());

a.path_segments_mut().map_err(|_| "").unwrap().push("a/b/../c");
println!("{}", a);
```
```bash
http://example.com/a/c
http://example.com/a%2Fb%2F..%2Fc
```

I think what you are suggesting could be implemented as a separate function, such as `add_path` or something similar.

davidhewitt replied:
Fair point, I had overlooked those possibilities 👍

davidhewitt reviewed (commented):
Thanks for the PR! The signature looks fine, though I think the implementation can be simpler.

davidhewitt commented on src/url.rs:
```diff
@@ -190,6 +190,37 @@ impl PyUrl {
+
+    #[pyo3(signature=(path, *args, no_trailing_slash=false))]
+    pub fn join(&self, path: &str, args: &Bound<'_, PyTuple>, no_trailing_slash: bool) -> PyResult<Self> {
+        let join_path = |base: &Url, partial_path: String| -> Result<Url, PyErr> {
+            base.join(&partial_path)
+                .map_err(|err| PyValueError::new_err(err.to_string()))
+        };
+
+        let add_trailing_slash = |path_str: &mut String| {
+            if !path_str.trim_end().ends_with('/') {
+                path_str.push('/');
+            }
+        };
+
+        let num_of_args = args.len();
+        let mut path_str = path.to_string();
+        if num_of_args != 0 || !no_trailing_slash {
+            add_trailing_slash(&mut path_str);
+        }
+
+        let mut new_url = join_path(&self.lib_url, path_str)?;
+
+        for i in 0..num_of_args {
+            let mut arg_str = args.get_item(i)?.to_string();
+            if i != num_of_args - 1 || !no_trailing_slash {
+                add_trailing_slash(&mut arg_str);
+            }
+            new_url = join_path(&new_url, arg_str)?;
+        }
```
I think this might be equivalent, reads simpler to me, and only does URL parsing once:

```suggestion
        let mut url = self.lib_url.to_string();
        
        let mut add_part = |part| {
            if !url.ends_with('/') {
                url.push('/');
            }
            url.push_str(part);
        }
        
        add_part(path);
        
        for path in args {
             add_part(path)
        }
        
        if !no_trailing_slash && !url.ends_with('/') {
             url.push('/');
        }
        
        let new_url = Url::parse(url).map_err(|err| PyValueError::new_err(err.to_string()));
```

... what do you think of this implementation?

Meetesh-Saini replied:
Yes this implementation looks better, but in the existing implementation I'm joining each and every part individually to support code like `a.join("/a/b?q=a","/c")` which will result in `http://example.com/c` while the suggested implementation will result in `http://example.com/a/b?q=a/c`.
I think this would be a fair assumption that user wants to join a continuous path and will have the fragments and queries at the end of the join. So this looks good to me.
I think there should be a check for path too.
```rust
// suggested
let mut add_part = |part| {
    if !url.ends_with('/') {
        url.push('/');
    }
    url.push_str(part);
}

// change to following
let mut add_part = |part : &str| {
    if !url.ends_with('/') {
        url.push('/');
    }

    if part.starts_with('/') {
        url.push_str(&part[1..]);
    } else {
        url.push_str(part);
    }
};
```
```bash
# for path: /a/b
# suggested will return
http://example.com//a/b
# while changed will return
http://example.com/a/b
```
Shall I commit this change?

davidhewitt replied:
This is a great question. In my opinion neither behaviour seems great, it seems surprising (to me) that passing an absolute url `/c` will cause all previous arguments to be disregarded. Are there precedents we can learn from (e.g. in `pathlib`, or maybe other URL handling libraries)?

Meetesh-Saini replied:
`pathlib` and `urllib.parse.urljoin` of Python and Rust's `url::Url.join` all show the same behavior where the absolute path overwrites the previous path of the URL. In my opinion, the best we can do is to provide a flag to the user `strict=false` which would join as the other libraries do, while the `strict=true` would join them as you suggested, considering the arguments strictly as part of the path.

Something similar to this,
```python
a = Url("https://example.com")
a.join("a", "/b") # https://example.com/b/
a.join("a", "/b", strict=True) # https://example.com/a/b/
```

Meetesh-Saini replied:
Hey @davidhewitt, just wanted to follow up on my last response

davidhewitt replied:
Sorry for the long delay; I do recall starting a response at some point but must have never finished it.

```python
>>> Path("foo").joinpath("a", "/b")
PosixPath('/b')
```

Ok, so Pathlib is variadic in the same way _and_ as you say does overwrite the base url. So your implementation looks correct to me and accounts for the complexities.

davidhewitt reviewed (commented):
I think this just needs test cases and then I would be happy to see this merged. Sorry for the long delay 😬

Meetesh-Saini commented:
I've added tests and changed the implementation. Now, it only takes one argument instead of multiple.
I noticed that using multiple arguments can be confusing, especially when `trailing_slash=False`. Keeping it to one argument makes things clearer, similar to how `servo-url` and `urllib.urljoin` work.
For example, the user may expect either of the following. I wasn't sure which one to implement.
```
            a.join("e", "?q=1", "g", trailing_slash=False) 
                    │     │      │                         
                    └──┬──┘      │                         
                       ▼         │                         
              ┌───► "e?q=1"      │                         
              │        │         │                         
    No trailing slash  └────┬────┘                         
                            ▼                              
                           "g"                             
                                                       
                                                                                       
         a.join("e", "?q=1", "g", trailing_slash=False)
                 │     │      │                        
                 └──┬──┘      │                        
                    ▼         │                        
                 "e/?q=1"     │                        
                    │         │                        
                    └────┬────┘                        
                         ▼                             
                       "e/g"                           
```

davidhewitt commented on python/pydantic_core/_pydantic_core.pyi:
```diff
@@ -606,6 +606,18 @@ class Url(SupportsAllComparisons):
             An instance of URL
         """
 
+    def join(self, path: str, trailing_slash: bool = True) -> Self:
```
Is `True` the right default here? Part of me thinks that if joining files like `a.png` (which might be quite common) that this default is unhelpful, especially as user can also get a trailing slash by passing it to the input.

Also, I think I would prefer the name `append_trailing_slash` as this doesn't strip them:

```suggestion
    def join(self, path: str, append_trailing_slash: bool = False) -> Self:
```

davidhewitt commented on src/url.rs:
```diff
@@ -190,6 +198,26 @@ impl PyUrl {
+            .join(path)
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+        if !trailing_slash || new_url.query().is_some() || new_url.fragment().is_some() || new_url.cannot_be_a_base() {
+            return Ok(PyUrl::new(new_url));
+        }
+
+        new_url
+            .path_segments_mut()
+            .map_err(|()| PyValueError::new_err("Url cannot be a base"))?
+            .pop_if_empty()
+            .push("");
```
I think the "cannot be a base" restriction seems unnecessary. How about this alternative:

```suggestion
        if trailing_slash && || !new_url.path().ends_with("/") && !(new_url.query().is_some() || new_url.fragment().is_some()) {
            match new_url.path_segments_mut() {
                Ok(segments) => segments.pop_if_empty().push(),
                Err(()) {
                    let mut new_path = new_url.path().to_string();
                    new_path.push('/');
                    new_url.set_path(&new_path);
                }
            }
        }
```

Meetesh-Saini replied:
Yes, I agree this alternative looks better.

davidhewitt reviewed (commented):
Sorry for yet another long review cycle. I would like to just agree what the right default for `trailing_slash` is, and have a suggestion to avoid the `cannot_be_a_base()` restriction.

davidhewitt commented on src/url.rs:
```diff
@@ -155,6 +155,14 @@ impl PyUrl {
         (self.__str__(),)
     }
 
+    fn __truediv__(&self, other: &str) -> PyResult<Self> {
+        self.join(other, true)
+    }
+
+    fn __floordiv__(&self, other: &str) -> PyResult<Self> {
+        self.join(other, false)
+    }
```
Ok, sorry I missed these in the last round of review. I think the difference between the `/` and `//` operators here is subtle and hard to document.

I think better we just have `/`, and make it so that it matches the default of `append_trailing_slash=False`. This will also simplify testing, I think.

```suggestion
    fn __truediv__(&self, other: &str) -> PyResult<Self> {
        self.join(other, false)
    }
```

davidhewitt replied:
Oh, I see. Yikes, there are so many subleties here!

It seems to me that our `.join()` method really works like `urllib.parse.urljoin` when it comes to semantics, e.g.

```python
>>> urllib.parse.urljoin("https://foo.com/a", "b")
'https://foo.com/b'
```

versus pathlib's

```python
>>> pathlib.Path("/foo/a").joinpath("b")
PosixPath('/foo/a/b')
```

Given these are inconsistent, I think we should perhaps back away from trying to have pathlib-like semantics at all.

Would you be open to the idea of dropping the operators from the PR completely, so we can get `.join()` merged? We could then open a `pydantic` issue to discuss the design of the operators and move forward with an implementation when there's consensus?

davidhewitt replied:
Alternatively we could also have `joinpath()` which works like `Pathlib` and doesn't accept query string or fragments as the whole input?

And then could have `/` operator work like `joinpath`? 🤔

Meetesh-Saini replied:
`joinpath()` would certainly make things cleaner. Should I implement `joinpath()` in this PR, or should we drop the operators for now and discuss it in the issues instead?

davidhewitt replied:
I think without comment from anyone else, let's just do `.join()` here and then follow-up with an issue in the main `pydantic` repo where we can discuss `.joinpath()` and operators.

Viicos replied:
I think it would be great to have more time to discuss the semantics (does it need to match `urllib`? What about other libraries like [`furl`](https://github.com/gruns/furl)? Should be double check with the current RFCs? We should also check what was said in [this discussion](https://discuss.python.org/t/31109)).

Meetesh-Saini replied:
> I think it would be great to have more time to discuss the semantics (does it need to match `urllib`? What about other libraries like [`furl`](https://github.com/gruns/furl)? Should be double check with the current RFCs? We should also check what was said in [this discussion](https://discuss.python.org/t/31109)).


Really sorry for the late response. The main URL joining part is handled by [rust-url's join method](https://docs.rs/url/latest/url/struct.Url.html#method.join) which implements the [WHATWG URL spec](https://url.spec.whatwg.org/). So, the `/` operator semantics are handled by the libraries in our case, python libraries like `urllib` and `furl`. While `furl` provides `/` operator, `urllib` does not. `urllib.parse.urljoin` and `furl.furl.join` follows the [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2.2) to resolve the new URL. 

`furl` is using `/` operator for only adding path to URL like [pathlib.Path](https://github.com/gruns/furl/blob/da386f68b8d077086c25adfd205a4c3d502c3012/README.md?plain=1#L290).

I think the `furl` approach is good. I should not have written my function signature as `signature=(path, append_trailing_slash=false)` but rather as `signature=(url, append_trailing_slash=false)` because the argument can be a relative or absolute url, not just path.

IMO, we can have `Url.join()` for URL joining, similar to `furl.furl.join`, and `Url.__truediv__` for the sole purpose of adding a path without a trailing slash, just like `furl.furl.__truediv__`.

davidhewitt reviewed (approved):
Thanks for the many iterations here and slow reviews by me. I am happy with the design here now. I would like to see the `__floordiv__` operator removed (see comment below), and then let's merge 👍

sydney-runkle commented:
> I think it would be great to have more time to discuss the semantics (does it need to match urllib? What about other libraries like [furl](https://github.com/gruns/furl)? Should be double check with the current RFCs? We should also check what was said in [this discussion](https://discuss.python.org/t/31109)).

Let's do this before moving forward
